### ISSUE TITLE
Provide the ability to generate a report from the search criteria

### ISSUE THREAD

**Issue by jadrake75** [stamp-web/stamp-web-aurelia#65]
The stamp content should be generatable as a PDF report.  Unlike previous generation it should match the table contents.

**Comment by jadrake75** [comment]
One thought I have is if we can generate a printable page then it could be saved as PDF

**Comment by zhhz** [comment]
On Mac, you can save any printable page as a PDF...

**Comment by jadrake75** [comment]
yep... that is exactly my current thinking.  If I can output valid css for
"printer" formats then I can simply send it to pdf and in theory it should
work for any display/table.

**Comment by jadrake75** [comment]
the end result should be a printable page that is a list of stamp information with a header block with information....

In my UI I am looking at a search results.... I have a table/grid of stamps.... you click the print

You get a dialog that shows "saved report formats" (a list/drop-down of them) or allows you to create a new one by providing a name and selecting from a series of "fields" and drag them into two blocks and then save it...

Header

Table columns

So the fields might be

•Country
•Description
•Active Catalogue Number
•Active Catalogue Value
•Condition
•Purchase Date

and may also include summary values like

•Total Count
•Total Catalogue Value

etc.

Effectively I want to make a report like

```
Country: China

Total Stamps: 256

Total Catalogue Value: $5460.75




Catalogue Number     Description     Value        Condition       Notes           

124a                 red (dragon)    $256.25      Mint            It is awesome

125                  green (dragon)  $1256.90     Mint    



```

So here is what I am thinking..... translate the output of the dialog above, and generate a html <template> that uses repeater over the stamps + is able to use my value-converters to convert "countryRef" into say "China" and then compile the template with Aurelia to generate the HTML!  I already have code now that I can pass HTML content to a utility service and it will generate a "fake" window, set the content and write it out along with a print.css and open the printer panel.....  I also have all the value converters etc. to write the stuff out in the correct way navigating the model so it comes down to building the template which I should be able to do from some sort of simple JSON like this:

```
{

   summary: [

      { "label": "Country i18n key", "value": "stamps[0].countryRef", "value-converter": "by-name" },

      { "label": "Total Stamps", "value": "stamps.total" },

      { "label": "Total Catalogue Value i18N key", "value": "stamps.catalogueTotal", "value-converter": "as-currency-formatted:'USD'"}

   ],

   table: [

     { "title": "Catalogue Number", "value": "stamps[?].activeCatalogueNumber.number"},

     { "title": "Description", "value": "stamps[?].description"},

     { "title": "Value", "value": "stamps[?].activeCatalogueNumber.value", "value-converter": "as-currency-formatted"}

     ...

   ]

}
```

I am using the syntax of [?] to represent the iterating stamp.  I would LOVE your feedback here!  Do you know of another syntax to use?  Maybe I can just use "stamp" knowing that it comes from "stamps" (which is the list of stamps to print)

this would generate something like

```
<template>
    <require from="../../value-converters/by-name">

   <require from="../../value-converters/as-currency-formatted">

    <div class="summary-panel">

       <label>Country:</label> ${stamps[0].countryRef | by-name: 'countryRef' }

       <label>Total Stamps:</label> ${stamps.total}

       <label>Total Catalogue Value:</label> ${stamps.catalogueTotal | as-currency-formatted:'USD'}

    </div>

    <div class="results-table">

       <table>

         <tr>

           <th repeat.for="header of table">${header.title}</th>

         </tr>

         <tr repeat.for="stamp of stamps">

             <!-- this part would have to be generated into the template -->

             <td>${stamp.activeCatalogueNumber.number}</td>

             <td>${stamp.description}</td>

             <td>${stamp.activeCatalogueNumber.value}</td>

         </tr>

       </table>

    </div>

</template>

```

which in theory I should be able to pass to a compile and generate the html for it and dump this as my content that will appear in my popup window for printing......

Kind of cool eh?  The other point I am thinking.... is the view format really is not print specific.  I "could" actually use that same "view" for display so instead of stamp-list.html having the table definition it is more like output the table based on generation - but that would be a phase 2.

Does this seem doable you think?  The fact I got the print working with a style sheet etc. is pretty cool.  I thought I could hard code the template for the first "iteration" to get something quick but I like this idea of dynamic table views.....
